relay/chunk_manager.go

```diff
@@ -0,0 +1,165 @@
+package relay
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	v2 "github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigenda/encoding"
+	"github.com/Layr-Labs/eigenda/encoding/rs"
+	"github.com/Layr-Labs/eigenda/relay/cache"
+	"github.com/Layr-Labs/eigenda/relay/chunkstore"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"golang.org/x/sync/errgroup"
+	"sync"
+	"sync/atomic"
+)
+
+type chunkManager struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// metadataCache is an LRU cache of blob metadata. Blobs that do not belong to one of the relay shards
+	// assigned to this server will not be in the cache.
+	frameCache cache.CachedAccessor[blobKeyWithMetadata, []*encoding.Frame]
+
+	// chunkReader is used to read chunks from the chunk store.
+	chunkReader chunkstore.ChunkReader
+
+	// pool is a work pool for managing concurrent requests. Used to limit the number of concurrent
+	// requests.
+	pool *errgroup.Group
+}
+
+// blobKeyWithMetadata attaches some additional metadata to a blobKey.
+type blobKeyWithMetadata struct {
+	blobKey  v2.BlobKey
+	metadata blobMetadata
+}
+
+func (m *blobKeyWithMetadata) Compare(other *blobKeyWithMetadata) int {
+	return bytes.Compare(m.blobKey[:], other.blobKey[:])
+}
+
+// newChunkManager creates a new chunkManager.
+func newChunkManager(
+	ctx context.Context,
+	logger logging.Logger,
+	chunkReader chunkstore.ChunkReader,
+	cacheSize int,
+	workPoolSize int) (*chunkManager, error) {
+
+	pool := &errgroup.Group{}
+	pool.SetLimit(workPoolSize)
+
+	server := &chunkManager{
+		ctx:         ctx,
+		logger:      logger,
+		chunkReader: chunkReader,
+		pool:        pool,
+	}
+
+	c, err := cache.NewCachedAccessor[blobKeyWithMetadata, []*encoding.Frame](cacheSize, server.fetchFrames)
+	if err != nil {
+		return nil, err
+	}
+	server.frameCache = c
+
+	return server, nil
+}
+
+// frameMap is a map of blob keys to frames.
+type frameMap map[v2.BlobKey][]*encoding.Frame
+
+// GetFrames retrieves the frames for a blob.
+func (s *chunkManager) GetFrames(ctx context.Context, mMap *metadataMap) (*frameMap, error) {
+
+	keys := make([]*blobKeyWithMetadata, 0, len(*mMap))
+	for k, v := range *mMap {
+		keys = append(keys, &blobKeyWithMetadata{blobKey: k, metadata: *v})
+	}
+
+	fMap := make(frameMap, len(keys))
+	hadError := atomic.Bool{}
+
+	wg := sync.WaitGroup{}
```

Comment:
same comment about the wg and errgroup

Reply:
Constructor changed. Let me know if you'd like to move away from the errgroup abstraction, and I'll change both locations.